File: companies/Trilogy/24.swap-nodes-in-pairs.java
/**
 * Algorithm:
 * Rewire pointers in pairs using a dummy predecessor. For pair first -> second,
 * reconnect prev -> second -> first -> nextPair.
 *
 * Complexity:
 * Time O(n), space O(1).
 */
class Solution {
    public ListNode swapPairs(ListNode head) {
        ListNode dummy = new ListNode(0, head);
        ListNode prev = dummy;
        while (prev.next != null && prev.next.next != null) {
            ListNode first = prev.next;
            ListNode second = first.next;
            first.next = second.next;
            second.next = first;
            prev.next = second;
            prev = first;
        }
        return dummy.next;
    }
}

File: companies/Meta/24.swap-nodes-in-pairs.cpp
#include <algorithm>
#include <cctype>
#include <climits>
#include <cmath>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <sstream>
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>
using namespace std;

// Definition for singly-linked list.
// struct ListNode {
//     int val;
//     ListNode *next;
//     ListNode() : val(0), next(nullptr) {}
//     ListNode(int x) : val(x), next(nullptr) {}
//     ListNode(int x, ListNode *next) : val(x), next(next) {}
// };


class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        /*
        Approach:
        Iteratively swap each adjacent pair by rewiring three next pointers. A
        dummy node gives a stable pointer before the current pair, including when
        the pair starts at the head.

        Complexity: O(n) time and O(1) space.
        */
        ListNode dummy(0, head);
        ListNode* prev = &dummy;
        while (prev->next && prev->next->next) {
            ListNode* first = prev->next;
            ListNode* second = first->next;
            first->next = second->next;
            second->next = first;
            prev->next = second;
            prev = first;
        }
        return dummy.next;
    }
};
